devices/usbc-cdc/ctrl.c

```diff
@@ -255,40 +277,57 @@ static int ctrl_buildDtd(dtd_t *dtd, u32 paddr, u32 size)
 
 dtd_t *ctrl_execTransfer(int endpt, u32 paddr, u32 sz, int dir)
 {
+	time_t start;
 	int shift;
 	u32 offs;
 	volatile dtd_t *dtd;
 
 	int qh = (endpt << 1) + dir;
 
-	dtd = ctrl_getDtd(endpt, dir);
+	if (ctrl_common.dc->connected) {
+		dtd = ctrl_getDtd(endpt, dir);
 
-	ctrl_buildDtd((dtd_t *)dtd, paddr, sz);
+		ctrl_buildDtd((dtd_t *)dtd, paddr, sz);
 
-	shift = endpt + ((qh & 1) ? 16 : 0);
-	offs = (u32)dtd & (((ctrl_common.dc->endptqh[qh].size) * sizeof(dtd_t)) - 1);
+		shift = endpt + ((qh & 1) ? 16 : 0);
+		offs = (u32)dtd & (((ctrl_common.dc->endptqh[qh].size) * sizeof(dtd_t)) - 1);
 
-	ctrl_common.dc->endptqh[qh].dtd_next = (ctrl_common.dc->endptqh[qh].base + offs) & ~1;
-	ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 6);
-	ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 7);
+		ctrl_common.dc->endptqh[qh].dtd_next = (ctrl_common.dc->endptqh[qh].base + offs) & ~1;
+		ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 6);
+		ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 7);
 
-	while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
-		;
+		while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
+			;
 
-	*(ctrl_common.dc->base + endptprime) |= 1 << shift;
+		*(ctrl_common.dc->base + endptprime) |= 1 << shift;
 
-	/* prime the endpoint and wait for it to prime */
-	while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
-		;
-	*(ctrl_common.dc->base + endptprime) |= 1 << shift;
-	while (!(*(ctrl_common.dc->base + endptprime) & (1 << shift)) && (*(ctrl_common.dc->base + endptstat) & (1 << shift)))
-		;
+		/* prime the endpoint and wait for it to prime */
+		while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
+			;
 
-	/* wait to finish transaction */
-	while (DTD_ACTIVE(dtd) && !DTD_ERROR(dtd))
-		;
+		*(ctrl_common.dc->base + endptprime) |= 1 << shift;
 
-	return (dtd_t *)dtd;
+		start = hal_timerGet();
+		while (!(*(ctrl_common.dc->base + endptprime) & (1 << shift)) && (*(ctrl_common.dc->base + endptstat) & (1 << shift)) && (*(ctrl_common.dc->base + portsc1) & 1)) {
+			if ((hal_timerGet() - start) > USBCTRL_TIMEOUT) {
+				return NULL;
+			}
+		}
+
+		/* wait to finish transaction while device is attached to the host */
+		while (DTD_ACTIVE(dtd) && !DTD_ERROR(dtd) && (*(ctrl_common.dc->base + portsc1) & 1)) {
+			if ((hal_timerGet() - start) > USBCTRL_TIMEOUT) {
+				return NULL;
+			}
+		}
```

Comment:
It will be a bit faster and smaller the result code.
```suggestion
		time_t end = hal_timerGet() + USBCTRL_TIMEOUT;
		while (!(*(ctrl_common.dc->base + endptprime) & (1 << shift)) && (*(ctrl_common.dc->base + endptstat) & (1 << shift)) && (*(ctrl_common.dc->base + portsc1) & 1)) {
			if (hal_timerGet() > end) {
				return NULL;
			}
		}

		/* wait to finish transaction while device is attached to the host */
		while (DTD_ACTIVE(dtd) && !DTD_ERROR(dtd) && (*(ctrl_common.dc->base + portsc1) & 1)) {
			if (hal_timerGet() > end) {
				return NULL;
			}
		}
```

Reply:
Thank you, change submitted.